src/v/archival/ntp_archiver_service.cc

```diff
@@ -2951,46 +2951,139 @@ size_t ntp_archiver::get_local_segment_size() const {
     return log_segment_size;
 }
 
-ss::future<bool>
-ntp_archiver::prepare_transfer_leadership(ss::lowres_clock::duration timeout) {
-    _paused = true;
-
+ss::future<bool> ntp_archiver::pause(ss::lowres_clock::duration timeout) {
     ss::gate::holder holder(_gate);
+
+    vlog(
+      _rtclog.info,
+      "Pausing archiver; active={}",
+      _uploads_active.current() != 0);
+
+    const auto deadline = ss::lowres_clock::now() + timeout;
+    bool success = true;
+
     try {
-        co_await ss::get_units(_uploads_active, 1, timeout);
+        // Phase 1: disable the housekeeping jobs and if the currently running
+        // job is managed by the archiver, then wait for it to complete
+        if (auto workflow_wrap = _parent.cloud_housekeeping_workflow();
+            workflow_wrap) {
+            if (_scrubber) {
+                _scrubber->set_enabled(false);
```

Comment:
I don't see any "mutexes" around this code. Is it possible that concurrently the scrubber  or merger to be re-activated due to a leadership change? aka can the `notify_leadership` method be aclled cocnurrently while we are suspended in one of the later co_awaits?

Reply:
Yeah, I think this is racy. `pause` and `unpause` should probably grab a mutex. Additionally, `notify_leadership` should be expressed in terms of `unpause`.